lib/active_merchant/billing/gateways/mundipagg.rb

```diff
@@ -0,0 +1,287 @@
+        '500' => STANDARD_ERROR_CODE[:processing_error]
+      }
+
+      STANDARD_ERROR_MESSAGE_MAPPING = {
+        '400' => 'Invalid request;',
+        '401' => 'Invalid API key;',
+        '404' => 'The requested resource does not exist;',
+        '412' => 'Valid parameters but request failed;',
+        '422' => 'Invalid parameters;',
+        '500' => 'An internal error occurred;'
+      }
+
+      def initialize(options={})
+        requires!(options, :api_key)
+        super
+      end
+
+      def purchase(money, payment, options={})
+        post = {}
+        add_invoice(post, money, options)
+        add_customer_data(post, options) unless payment.is_a?(String)
+        add_shipping_address(post, options)
+        add_payment(post, payment, options)
+
+        commit('sale', post)
+      end
+
+      def authorize(money, payment, options={})
+        post = {}
+        add_invoice(post, money, options)
+        add_customer_data(post, options) unless payment.is_a?(String)
+        add_shipping_address(post, options)
+        add_payment(post, payment, options)
+        add_capture_flag(post, payment)
+        commit('authonly', post)
+      end
+
+      def capture(money, authorization, options={})
+        post = {}
+        post[:code] = authorization
+        add_invoice(post, money, options)
+        commit('capture', post, authorization)
+      end
+
+      def refund(money, authorization, options={})
+        add_invoice(post={}, money, options)
+        commit('refund', post, authorization)
+      end
+
+      def void(authorization, options={})
+        commit('void', post=nil, authorization)
+      end
+
+      def store(payment, options={})
+        post = {}
+        options.update(name: payment.name)
+        options = add_customer(options) unless options[:customer_id]
+        add_payment(post, payment, options)
+        commit('store', post, options[:customer_id])
+      end
+
+      def create_customer(options={})
+        post = {}
+        post[:name] = options[:name]
+        commit('customer', post)
+      end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript
+          .gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]')
+          .gsub(%r(("cvv\\":\\")\d*), '\1[FILTERED]')
+          .gsub(%r((card\\":{\\"number\\":\\")\d*), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_customer(options)
+        customer = create_customer(options)
+        options.update(customer_id: customer.authorization)
+      end
+
+      def add_customer_data(post, options)
+        post[:customer] = {}
+        post[:customer][:email] = options[:email]
+      end
+
+      def build_billing_address(options)
```

Comment:
For consistency sake, name this method `add_billing_address`?

Reply:
So I had it that way and changed it since its "building" the billing address object and not really "adding" it. Maybe I am being too literal.